### PR TITLE
Adds Cassandra client and server integration

### PR DESCRIPTION
**Pull request by adriancole**
This contains tracing instrumentation for [Cassandra](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/tracing/Tracing.java) and the [DataStax Java Driver](https://github.com/datastax/java-driver).

`brave.cassandra.Tracing` extracts trace state from the custom payload
of incoming requests. How long each request takes, each suboperation,
and relevant tags like the session ID are reported to Zipkin.

`brave.cassandra.driver.TracingSession` tracks the client-side of cassandra and
adds trace context to the custom payload of outgoing requests. If
server integration is in place, cassandra will contribute data to these 
RPC spans.

### CODE REVIEW

File: instrumentation/cassandra/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Clients must enable tracing. If they want to continue an existing trace,
+they must add propagation fields to the custom payload of a statement.
+
+```java
+// minimally, you need to prepare a statement and enable tracing
+preparedStatement.enableTracing();
+
+// By default, B3 style is used, so instrumented clients do something like this
+Map<String, ByteBuffer> payload = new LinkedHashMap<>();
+payload.set("X-B3-TraceId", byteBuffer("463ac35c9f6413ad"));
```
Comment (adriancole) [inline instrumentation/cassandra/README.md]: note that brave allows you to change the propagation format. I intentionally decided to not go with the finagle binary encoding as most people don't understand what that is. B3 is reasonably well documented, so a better default IMHO.

**Comment by adriancole** [comment]
PS I will raise another pull request that depends on this for datastax client shortly (harvesting code from zipkin and some buffing)

**Comment by adriancole** [comment]
Ok added the client side to this pull request. I have some test cases to fix, and need to put in basic tagging policy. Should be ready in the next day or two.

**Comment by adriancole** [comment]
I got all the normal tests working. So, folks can test this (I'll also change zipkin to use it).

Next steps are to put in some basic tag policy logic, and add edge-case tests we don't usually cover, like write timeouts.

**Comment by adriancole** [comment]
added tag policy hooks to the client side (the only code zipkin needs at the moment)

still TODO: edge case tests and server-side policy

**Comment by adriancole** [comment]
buffed the `CassandraDriverParser`, so it looks prettier and accepts Tagger not Span

**Comment by adriancole** [comment]
I'm going to rename CassandraDriverX to CassandraClientX in a separate commit. @llinder etc let me know if you disagree with this choice.

**Comment by adriancole** [comment]
OK I think this is good for a first cut. I'll integrate in zipkin and if it works have no more changes planned for the initial impl. We can always add more functionality later (and likely will need to)

**Comment by adriancole** [comment]
updated to use the SpanCustomizer type

**Comment by adriancole** [comment]
I'm not going to merge this until some folks savvy with cassandra have a chance to play around with it.

**Comment by adriancole** [comment]
ps considering the scope of this includes 3 modules, it might make more sense to bump this to a separate repository brave-cassandra. Any thoughts?

**Comment by devinsba** [comment]
:+1: to new repo if there will be a bunch of churn on that. Not sure what the cadence is for the cassandra project

**Comment by llinder** [comment]
Separate repository sounds good to me.

**Comment by llinder** [comment]
Looks like a solid start.  I will give this a go in one of our services in pre-production as soon as I can.  Just want to finish up the Ratpack work first.

File: instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    });
+    version = delegate.getCluster().getConfiguration().getProtocolOptions().getProtocolVersion();
+  }
+
+  @Override public ResultSetFuture executeAsync(Statement statement) {
+    Span span = nextSpan(statement);
+    if (!span.isNoop()) parser.request(statement, span.kind(CLIENT));
+
+    // o.a.c.tracing.Tracing.newSession must use the same propagation format
+    if (version.compareTo(ProtocolVersion.V4) >= 0) {
```
Comment (michaelsembwever) [inline instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java]: does it break the request if outgoingPayload is added to it against a server that doesn't support it (ie < V4) ?

File: instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    ResultSetFuture result;
+    try {
+      result = delegate.executeAsync(statement);
+    } catch (RuntimeException | Error e) {
+      if (span.isNoop()) throw e;
+      parser.error(e, span);
+      span.finish();
+      throw e;
+    }
+    Futures.addCallback(result, new FutureCallback<ResultSet>() {
```
Comment (michaelsembwever) [inline instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java]: have we any idea on the impact adding all the callbacks will cause.
EG in a client that processes 10k+ requests per second this could cause trouble, and a suitable load-shedding mechanism be appropriate.

File: instrumentation/cassandra/src/main/java/brave/cassandra/CassandraTraceKeys.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package brave.cassandra;
+
+public final class CassandraTraceKeys {
```
Comment (michaelsembwever) [inline instrumentation/cassandra/src/main/java/brave/cassandra/CassandraTraceKeys.java]: it's interesting that you use `final class` here, as appropriate, yet not on the other class.
i would not presume that the other classes are designed for being subclassed (which are in fact broadening the API being published beyond what you're probably intending it to be.)

**Comment by adriancole** [comment]
>
> +public final class CassandraTraceKeys {
>
> it's interesting that you use final class here, as appropriate, yet not
> on the other class.
> i would not presume that the other classes are designed for being
> subclassed (which are in fact broadening the API being published beyond
> what you're probably intending it to be.)
>
This is just a habit for constant files, i wouldn't read too much into
that. If someone were subclassing, they could I guess make a subclass of
TraceKeys.. wasn't thinking about that.

The other class is subclassable mainly as it seems very likely to need it.
I mentioned some notes in the type doc about configuration maybe being
something not awesome with system properties. Another thing is how
subordinate work is recorded.. right now, for simplicity and also to keep
traces from getting too "bottom heavy", I put annotations instead of child
spans. Subclassing would allow someone to override that, and since it is
likely they'll need to override the construction of the type anyway (bc it
is literally new'ed) seemed to fit.

I suspect this code will flex a bit before final, but there were my
thoughts going into it..

File: instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java (hunk 3 above)
Comment (adriancole) [inline instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java]: heh actually looking at the code, this is a bit silly.. I am checking span.isNoop for each callback. Will buff this out and move the conditional in front. I don't think we'd want to sample 10K spans/second, so this should take care of that concern.

Anyway, you highlight something which is that we don't yet have benchmarks on this for normal vs instrumented case.

File: instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java (hunk 2 above)
Comment (adriancole) [inline instrumentation/cassandra-driver/src/main/java/brave/cassandra/driver/TracingSession.java]: you'd know better than me :) To test this though, I suppose we could add an invoker test that uses an older version of cassandra..

**Comment by adriancole** [comment]
Tracing in Cassandra is actively maintained, ex there are some notes about
things that work in 3.11. I suspect it will actually be the testing and
focus concerns that make this more sensible as a separate repo beyond just
the several modules reason.

For example, we will likely need a matrix build at some point albeit via
maven invoker or docker. This is easier as a separate repo.

On focus, there is a lot of activity in brave unrelated to cassandra.
Another repo could keep the spam level down a bit.

File: instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    return contextOrFlags.context() != null
+        ? tracer.joinSpan(contextOrFlags.context())
+        : tracer.newTrace(contextOrFlags.samplingFlags());
+  }
+
+  @Override protected final void stopSessionImpl() {
+    ZipkinTraceState state = (ZipkinTraceState) get();
+    if (state != null) state.incoming.finish();
+  }
+
```
Comment (michaelsembwever) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: how come you didn't also override `public void doneWithLocalSession(TraceState)` ?

Comment (adriancole) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: possibly an oversight?

File: instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  protected void parseRequest(TraceState state, String request, Map<String, String> parameters,
+      SpanCustomizer customizer) {
+    customizer.name(parseSpanName(state, request));
+    customizer.tag(CassandraTraceKeys.CASSANDRA_REQUEST, request);
+    customizer.tag(CassandraTraceKeys.CASSANDRA_SESSION_ID, state.sessionId.toString());
+  }
+
+  @Override protected final TraceState newTraceState(InetAddress coordinator, UUID sessionId,
+      TraceType traceType) {
+    throw new AssertionError();
```
Comment (michaelsembwever) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: why is this an AssertionError?

**Comment by adriancole** [comment]
I've moved all this code here for future updates. feel free to adjust anything! https://github.com/openzipkin/brave-cassandra

File: instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java (hunk 5 above)
Comment (michaelsembwever) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: Or `public TraceState initializeFromMessage(final MessageIn<?> message)` ?
Or `public Map<String, byte[]> getTraceHeaders()`

(by not implementing outbound tracing, you won't be getting the tracing across the nodes *within* the cluster)

File: instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java (hunk 6 above)
Comment (adriancole) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: because this method shouldn't be called because we override what calls it. That said, if the upstream code drifts that could cease to be the case.

File: instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java (hunk 5 above)
Comment (adriancole) [inline instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java]: honestly I only implemented barebones to get things off the ground. Notably there's a lot of moving parts just in basics, so I was hoping to get something together that people can contribute to to add more features.

**Comment by adriancole** [comment]
added https://github.com/openzipkin/brave-cassandra/issues/1 to not lose
this point

**Comment by adriancole** [comment]
ok! the brave-cassandra repo is in business. For example, you can grab the snapshot cassandra jar here http://oss.jfrog.org/oss-snapshot-local/io/zipkin/brave/cassandra/brave-instrumentation-cassandra/0.0.1-SNAPSHOT/

Thanks for the review folks, and please direct all future things here, por favor! https://github.com/openzipkin/brave-cassandra
